File: dynamic_knapsack.py
prices = {}
prices["stereo"] = 3000
prices["laptop"] = 2000
prices["guitar"] = 1500
prices["iPhone"] = 2000
# prices["mp3"] = 1000

weights = {}
weights["stereo"] = 4
weights["laptop"] = 3
weights["guitar"] = 1
weights["iPhone"] = 1
# weights["mp3"] = 1

capacity = 4

items = list(prices.keys())

lightest = float("inf")
for weight in weights.values():
  if weight < lightest:
    lightest = weight

num_cols = int(capacity / lightest)

print(num_cols)
print(items)

knapsack = []
for row in range(len(items)):
  knapsack.append([])
  for col in range(num_cols):
    knapsack[row].append({})
    sub_capacity = (col * lightest) + lightest
    if row == 0:
      cell = {}
      cell["items"] = []
      cell["value"] = 0
      if weights[items[row]] <= sub_capacity:
        cell["items"] = [items[row]]
        cell["value"] = prices[items[row]]
      knapsack[row][col] = cell
    if row > 0:
      prev_max = knapsack[row-1][col]
      prev_max_value = 0
      if "value" in prev_max.keys():
        prev_max_value = prev_max["value"]
      this_value = 0
      topup_value = 0
      if weights[items[row]] <= sub_capacity:
        topup_amount = sub_capacity - weights[items[row]]
        if topup_amount > 0:
          topup_value = knapsack[row-1][int(topup_amount / lightest) - 1]["value"]
        current_item_value = prices[items[row]]
        this_value = topup_value + current_item_value
      if this_value > prev_max_value:
        cell = {}
        cell["items"] = [items[row]]
        cell["value"] = prices[items[row]]
        if topup_value > 0:
          cell["items"] += knapsack[row-1][int(topup_amount / lightest) - 1]["items"]
          cell["value"] += topup_value
        knapsack[row][col] = cell
      else:
        knapsack[row][col] = knapsack[row-1][col]

for row in knapsack:
  print(row)

print("Optimal item pick =", knapsack[len(knapsack)-1][num_cols - 1])

File: dynamicKnapsack.js
const items = [
  {name: "guitar", weight: 1, price: 1500},
  {name: "stereo", weight: 4, price: 3000},
  {name: "laptop", weight: 3, price: 2000},
  {name: "iPhone", weight: 1, price: 2000},
  {name: "mp3", weight: 1, price: 1000}
];

const capacity = 4 // knapsack capacity

let increment = Number.POSITIVE_INFINITY;
items.forEach(item => {
  if (item.weight < increment) {
    increment = item.weight;
  }
});

const numCols = parseInt(capacity / increment);

const knapsack = [];
items.forEach((item, row) => {
  knapsack.push([]);
  for (let col = 0; col < numCols; col++) {
    knapsack[row].push([]);
    const colMaxCapacity = (col * increment) + increment;
    if (row === 0) {
      if (item.weight <= colMaxCapacity) {
        knapsack[row][col].push(item);
      } else {
        knapsack[row][col].push({name: None, price: 0});
      }
    } else {
      const prevMax = knapsack[row-1][col];
      const prevMaxValue = prevMax.reduce((acc, pmItem) => {
        return acc + pmItem.price;
      }, 0);
      if (item.weight <= colMaxCapacity) {
        const addnlCapacity = colMaxCapacity - item.weight;
        if (addnlCapacity > 0) {
          const addnlCapacityValue = knapsack[row - 1][parseInt(addnlCapacity / increment) - 1]
            .reduce((acc, acItem) => {
            return acc + acItem.price;
          }, 0);
          if ((addnlCapacityValue + item.price) > prevMaxValue) {
            knapsack[row][col].push(item);
            knapsack[row - 1][parseInt(addnlCapacity / increment) - 1]
              .forEach(addnlItem => knapsack[row][col].push(addnlItem));
          } else {
            knapsack[row][col] = prevMax;
          }
        } else {
          if (item.price > prevMaxValue) {
            knapsack[row][col].push(item);
          } else {
            knapsack[row][col] = prevMax;
          }
        }
      } else {
        knapsack[row][col] = prevMax;
      }
    }
  }
});

console.log("\nOPTIMAL COMBINATION\n-------------------");
const optimalCombination = knapsack[knapsack.length - 1][numCols - 1];
let totalValue = 0;
optimalCombination.forEach(item => {
  totalValue += item.price;
  console.log(item.name, "\t", item.price)
});
console.log("Total\t", totalValue);
